src/sql/platform/connection/common/providerConnectionInfo.ts

```diff
@@ -74,12 +75,22 @@ export class ProviderConnectionInfo extends Disposable implements azdata.Connect
 			if (capabilities) {
 				this._serverCapabilities = capabilities.connection;
 			}
-			this._register(this.capabilitiesService.onCapabilitiesRegistered(e => {
+			if (this._onCapabilitiesRegisteredDisposable) {
+				dispose(this._onCapabilitiesRegisteredDisposable);
+			}
+			this._onCapabilitiesRegisteredDisposable = this.capabilitiesService.onCapabilitiesRegistered(e => {
```

Comment:
Per my email, I do wonder if we need to register this in the case where the capabilities cache is completed (e.g. extensions completed loading & capabilities has your provider, so it's guaranteed "fresh").

Also does this PR cover the scenario where ConnectionDialog creates 1 connection per object in recent connections, and then these all get leaked on close?

Reply:
I think we still would in case more capabilities are loaded later - for example if an extension wasn't loaded on launch and instead on an event or something and then registered a capability at that point or if it was disabled/re-enabled. 

We could also change it so that instead of caching the capabilities we query the service each time and just get the latest capabilities at that point.

@anthonydresser @kburtram Either of you have opinions on which approach we should take? Disposing the connections properly is still something we should do I think - although as mentioned given how much these are passed around I'm thinking the best approach is just to try and make this not require any special cleanup since otherwise it's highly likely we'll continue to miss cleaning these up and need to keep fixing that in the future.

Reply:
@kevcunnane Just looked into it and I found at least one way we're not correctly disposing the ones made in the recent connections list so I'll go fix that up now as well. Thanks for spotting that!

Reply:
Yeah and just to be clear, my thinking is that "loaded later" can be safely ignored in some cases:
- If the all (currently installed) extensions are loaded, and you've found existing capabilities earlier in the method, then these are definitely correct. The extension that provides the capabilities is fully loaded so we don't need a listener
- If extensions are loaded and you did not find the capabilities, you need a listener to handle someone reinstalling the relevant extension. In this case, still need listener.

If we wait on `extensionService.whenInstalledExtensionsRegistered()` promise, then re-assess the capabilities we can cover both cases:
- After extensions load, if we find the capabilities we are "done". That's 100% the correct values
- If we do not find capabilities, the extension providing it isn't installed. Add the listener as we have no other option.

The capabilities come from package.json so they'll have been updated during extension initialization.

Reply:
What about if someone installs a new extension after initial load that registers new capabilities for an existing provider ID?

Although would that even be a valid scenario we want to support? Shouldn't provider IDs be completely unique? I don't see us explicitly preventing that which seems like something we should consider changing.

Reply:
There also seems to be a bug right now - if an extension with a provider is uninstalled then the connection profiles never get updated and so continue thinking that the provider has those capabilities even though it isn't around anymore. Although at that point the whole connection profile is kind of broken anyways so maybe that isn't a scenario we have to be super concerned about at this level.